### PR TITLE
Add upgrade plugin command

### PR DESCRIPTION
**Pull request by asvechnikov**
Add upgrade plugins command to octane.
          octane update-plugin-settings --plugins plugin1,plugin2 ORIG_ID SEED_ID
This command should transfer settings for specified plugins from original cluster to seed cluster.

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -35,3 +35,4 @@ octane =
     upgrade-db = octane.commands.upgrade_db:UpgradeDBCommand
     install-node = octane.commands.install_node:InstallNodeCommand
     upgrade-control = octane.commands.upgrade_controlplane:UpgradeControlPlaneCommand
+    upgrade-plugin = octane.commands.upgrade_plugin:UpgradePluginCommand
```
Comment (akscram) [inline setup.cfg]: Rename to `upgrade-plugins`.

File: octane/commands/upgrade_plugin.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+            type=int,
+            metavar='ORIG_ID',
+            help="ID of original environment")
+        parser.add_argument(
+            'seed_env',
+            type=int,
+            metavar='SEED_ID',
+            help="ID of seed environment")
+        parser.add_argument(
+            'plugins',
```
Comment (akscram) [inline octane/commands/upgrade_plugin.py]: It should be used in singular because it should be repeated like this (I've wrote about it in the previous comments).

    upgrade-plugins PLUGINS [PLUGINS ...] # <- It's not correct.

I'm not sure to use metavars at all. Why did you add it?

File: requirements.txt [hunk 3]
```diff
@@ -16,3 +16,5 @@ python-keystoneclient<=0.11.1  # the last version without too fresh requirements
 python-fuelclient>=6.1
 cliff>=1.7.0,<=1.9.0  # should already be pulled by python-fuelclient
 paramiko==1.13.0
+
+zabbix-client
```
Comment (akscram) [inline requirements.txt]: zabbix-client==0.1.0

File: octane/commands/upgrade_plugins.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+            'emc_username': {'value': astute['storage']['emc_username']},
+            'emc_pool_name': {'value': astute['storage']['emc_pool_name']},
+            'metadata': {'enabled': astute['storage']['volumes_emc']}}
+
+
+def zabbix_snmptrapd_settings(env):
+    # go to controller and parse /etc/snmp/snmptrapd.conf
+    # ACCESS CONTROL sections
+    node = next(get_controllers(env))
+    data, _ = ssh.call(['cat', '/etc/snmp/snmptrapd.conf'], stdout=ssh.PIPE, node=node)
```
Comment (YorikSar) [inline octane/commands/upgrade_plugins.py]: You should instead do:

    with ssh.sftp(node).open('/etc/snmp/snmptrapd.conf') as f:
        data = f.read()

File: setup.cfg [hunk 5]
```diff
@@ -35,3 +35,4 @@ octane =
     upgrade-db = octane.commands.upgrade_db:UpgradeDBCommand
     install-node = octane.commands.install_node:InstallNodeCommand
     upgrade-control = octane.commands.upgrade_controlplane:UpgradeControlPlaneCommand
+    upgrade-plugins = octane.commands.upgrade_plugins:UpgradePluginsCommand
```
Comment (YorikSar) [inline setup.cfg]: Let's name it `update-plugin-settings`

File: octane/commands/upgrade_plugins.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+            'metadata': {'enabled': True}}
+
+
+def transfer_plugins_settings(orig_env_id, seed_env_id, plugins):
+    orig_env = environment.Environment(orig_env_id)
+    seed_env = environment.Environment(seed_env_id)
+    attrs = {}
+    for plugin_name in plugins:
+        plugin = plugin_name.replace('-', '_')
+        attrs[plugin] = PLUGINS[plugin_name](orig_env)
```
Comment (YorikSar) [inline octane/commands/upgrade_plugins.py]: Every plugin should receive only parsed `astute.yaml` and return a dict.

File: octane/commands/upgrade_plugins.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+        print(plugin)
```
Comment (YorikSar) [inline octane/commands/upgrade_plugins.py]: You should do smth like `LOG.info("Fetching settings for plugin '%s'", plugin)` before calling plugin instead. `print` is evil here.

File: octane/commands/upgrade_plugins.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+            'seed_env',
+            type=int,
+            metavar='SEED_ID',
+            help="ID of seed environment")
+        parser.add_argument(
+            'plugins',
+            metavar='PLUGIN',
+            choices=PLUGINS,
+            nargs="+",
+            help="Choose from {0}".format(PLUGINS.keys()))
```
Comment (YorikSar) [inline octane/commands/upgrade_plugins.py]: It should be an option, not a positional argument. Let's do it like `octane update-plugin-settings --plugins a,b,c 1 2`.

To do so, you should pass a verification and conversion function as `type` argument to `add_argument` like this:

    def plugin_names(s):
        plugins = s.split(',')
        for plugin in plugins:
            if plugin not in PLUGINS:
                raise TypeError("Unknown plugin {0}".format(plugin))
        return plugins

File: octane/commands/upgrade_plugins.py (hunk 7 above)
Comment (YorikSar) [inline octane/commands/upgrade_plugins.py]: You should do smth like `LOG.info('Fetching attrs for plugin %s', plugin)` before calling out to plugin instead. Plain `print` is evil.

**Comment by akscram** [comment]
Your changes do not pass some of PEP8 checks, use `tox -e pep8` command to identify and fix them.

File: octane/commands/update_plugin_settings.py [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+        data = f.read()
+    template = re.compile(r"authCommunity\s[a-z-,]+\s([a-z-]+)")
+    match = template.search(data)
+    return {'community': {'value': match.group(1)},
+            'metadata': {'enabled': True}}
+
+def get_zabbix_client(astute):
+    url = get_zabbix_url(astute)
+    user, password = get_zabbix_credentials(astute)
+    session = requests.Session();
```
Comment (akscram) [inline octane/commands/update_plugin_settings.py]: There is a semicolon at the end of the line. Keep the next lines closer together:

    session = requests.Session()
    session.proxies = {
        ...
    }

**Comment by YorikSar** [comment]
Please add test covering at least argument parsing and replace `':'.join((host['name'], host['ip']))` constructs with `"{0}:{1}".format(host['name'], host['ip'])`

**Comment by YorikSar** [comment]
Looks fine. Someone should merge it with up-to-date `telco` branch and verify that it still works.
